bin/ch-completion.bash

```diff
@@ -0,0 +1,403 @@
+        fi
+        ;;
+    undelete )
+        # FIXME: Update with “ch-image undelete --list” once #1551 drops
+        COMPREPLY=()
+        ;;
+    '' )
+        # Only autocomplete subcommands if there's no subcommand present.
+        COMPREPLY=( $(compgen -W "$_image_subcommands" -- "$cur") )
+        ;;
+    esac
+
+    # If we've made it this far, the last remaining option for completion is
+    # common opts. Note that we do the “-n” check to avoid being overzealous
+    # with our suggestions.
+    if [[ -n "$cur" ]]; then
+        COMPREPLY+=( $(compgen -W "$_image_common_opts" -- "$cur") )
+        return 0
+    fi
+}
+
+## ch-run ##
+
+# Completion function for ch-run
+#
+_ch_run_completion () {
+    echo "does nothing"
+}
+
+## Helper functions ##
+
+# Disable completion.
+ch-completion-disable () {
+    complete -r ch-image
+}
+
+# Figure out which storage directory to use (including cli-specified storage).
+_ch_find_storage () {
+    if echo "$@" | grep -Eq -- '\s(--storage|-\w*s)'; then
+        # This sed only works as desired if “--storage” or “-s” are present in
+        # the command line...
+        sed -E 's/(.*)(--storage=*|[^-]-s=*)\ *([^ ]*)(.*$)/\3/g' <<< "$@"
+    elif [[ -n "$CH_IMAGE_STORAGE" ]]; then
+        echo "$CH_IMAGE_STORAGE"
+    else
+        echo "/var/tmp/$USER.ch/"
+    fi
+}
+
+# Horrible, disgusting function to find an image or image ref in the ch-run
+# command line.
+#
+# NOT FINISHED, DON'T USE!!!
+_ch_run_image_finder () {
+    # Takes array of words. Tries to find an image in there.
+    #local -n words=$1
+    local wrds=("$@")
+    local ct=1
+
+    #echo "len: ${#wrds[@]}"
+
+    while ((ct < ${#wrds[@]})); do
+        if [[    (    -f ${wrds[$ct]} \
+                   && (    ${wrds[$ct]} == *.sqfs \
+                        || ${wrds[$ct]} == *.tar.? \
+                        || ${wrds[$ct]} == *.tar.?? \
+                        || ${wrds[$ct]} == *.tgz ) ) \
+              || (    -d ${wrds[$ct]} \
+                   && ${wrds[$ct-1]} != --mount \
+                   && ${wrds[$ct-1]} != -m \
+                   && ${wrds[$ct-1]} != --bind \
+                   && ${wrds[$ct-1]} != -b ) ]]; then
+            echo "${wrds[$ct]}"
+        fi
+        # FIXME: Check for image refs
+        ((ct++))
+    done
+}
+
+# Print the subcommand in an array of words; if there is not one, print an empty
+# string. This feels a bit kludge-y, but it's the best I could come up with.
+# It's worth noting that the double for loop doesn't take that much time, since
+# the Charliecloud command line, even in the wost case, is relatively short.
+#
+# Usage: _ch_image_subcmd_get [words]
+#
+# Example:
+#   >> _ch_image_subcmd_get "ch-image [...] build [...]"
+#      build
+_ch_image_subcmd_get () {
+    local cmd
+    local subcmd=
+    for word in "$@"
+    do
+        for cmd in $_image_subcommands
+        do
+            if [[ "$word" == "$cmd" ]]; then
+                subcmd="$cmd"
```

Comment:
Don't `cmd` and `subcmd` both refer to a subcommand? If so, the naming of `cmd` is confusing.